### ISSUE TITLE
Add ability to run an arbitrary command prior to connection.

### ISSUE THREAD

**Issue by BrainwreckedTech** [FreeRDP/Remmina#520]
This is a request to add a feature that would allow someone to run an arbitrary command prior to establishing a connection.  This could be added to the Advanced Tab in the Remote Desktop Preferences dialog.  To keep it simple, Remmina shouldn't concerned itself with the command after issuing it.  The command should be forked, allowing the command to run in the background and allowing Remmina to continue its business.  

This could be used, for example, to start a QEMU instance on the local machine.  This would make the process of starting a VM and connecting to it via VNC a one-stop process.

**Comment by antenore** [comment]
This is a nice one, I vote for it, I'm not sure we will have the time to implement it ASAP.
Personally I'm very slow (I mean it), so it could take really ages...

Are you up to it? Or do you know someone that could be interested? We don't bite, blame or critics...

**Comment by BrainwreckedTech** [comment]
Hmmmm.....well, TBH, I'm going to be slow at it as well.

TL;DR: By the time I can get to it, a beginning college student could have done it, even if that student had zero experience with C and Glade and had to look up every single step needed in both.  It does seem small enough to take on, though.

1) I've never done much outside of Visual Basic, and very little with C.  (So in a way, even at my age, I'm going to be that aforementioned college student.)  Counterpoint: I recently launched qemu-launcher on github, so some initial ass-kicking with GUIs is out of the way.  Even though I went with Python2/wxFormBuilder, I did play with Glade.  At least Remmina is in C and not C++ (thanks, Linus, for confirming that I wasn't alone in having a slight hatred for C++), and it's not like I've _never_ used C before.

2) I'm going to be having back surgery this week.  Not a good time to start things.  The fun begins Wednesday, the surgery is on Friday, an expected 4-day hospital stay, and an expected flurry of check-ups, rehab, and paperwork for at least 6 weeks.   Problem started before Christmas, but there was a delay in getting insurance.  (It started out as a 30-day response time, and that was missed.  Then it was 60 days, and that was missed.  It ended up taking 73 days to get insurance.) If the original 30-day period had been met, I probably would have had surgery the day before Christmas as that's when the pain got so bad my legs gave out on me.

3) My other github project, aurpb, is LONG overdue for an update due to Arch Linux removing --asroot from pacman.  I never fully tested my changes out of (misplaced) hope that I'd be admitted to the hospital.  If I had known then what I know now, it would have gone ahead and done what needed to be done.  Testing the changes won't take _that_ long, I just didn't want to get into a marathon, have an admission in the middle of it, and then come back completely lost.

**Comment by antenore** [comment]
I'm really sorry for your current situation...

Regarding your request, don't forget that you can execute remmina from the command line.

    remmina -c ~/.remmina/1405462424144.remmina 
 
Therefore you can write a script that handle the qemu and than remmina connection for you.

Beside this, I find your request quite interesting as it can be really useful also in other situations.
One of us was developing something similar (the other way around) and I think I'm going to reuse part of his code.

Tomorrow I'll let you know.

**Comment by antenore** [comment]
@BrainwreckedTech hope you are going fine.

Can you test and let me know? It's still a preview and has some minor bugs, but can be used.

**Comment by BrainwreckedTech** [comment]
A little better every day.  Getting around without a walker.  Able to stand up and sit down more smoothly.  Still have the same partial feeling in my legs, though.  And I can't be in the standard using-computer position for more than an hour.

The only problem is that VNC connection is attempted immediately after *and* is given no timeout.  This results in an immediately failure for the VNC connection.  The command execution part works as expected, though.

Command: mate-terminal -e "sudo qcl -w -x 50 -m 512 -p /mnt/qvm/clib-arch.img"

Even if qcl is added to NOPASSWD in sudoers, the VNC connection still fails too quickly.

**Comment by antenore** [comment]
Yes, this is annoying, I've to improve the code so that it waits before to open the connection.

There are different possibilities:

1. An "in progress" dialog box  (suggested by @giox069 )

This is maybe the cleanest. The command can be canceled if needed, plus, eventually a timeout

2. A global timeout

The command is interrupted after a fixed, global timeout.
This solution should be implemented anyway.

3. Per connection timeout

Probably it would be a good idea to be able to have more options and the possibility to decide when to execute a command (before, during, after) or to set different commands per each connection step (before, during and again after). Each of this option with a predefined and a connection, specific, timeout.

So finally, probably, all of the above should be implemented.

**Comment by BrainwreckedTech** [comment]
Is executing the command in the foreground a good idea, or am I getting the wrong impression?  My interpretation of "foreground execution" is that the command must finish execution before attempting the VNC connection.  This won't work for `qcl` nor for `qemu` itself, unless the command line can be something like

```
mate-terminal -e "sudo qcl [options]" & sleep 2
```

which would (hopefully) allow `mate-terminal` to execute in the background while `sleep` forces the wait needed for `qemu` to get up and running.

**Comment by antenore** [comment]
It depends by the scenario.

In general I would say yes and it's the main reason of this new feature.

In my case, having the script/program finished and executed successfully, it's a need and the same it's for you.

Some users may need to execute a command before the connection but in background, therefore we should implement also a foreground/background flag.

Now, my code miss completely the foreground part :-P .I'm working to fix it (as usual in my limited free time).

**Comment by giox069** [comment]
The backgroud execution can just be solved externally running a script which launches a backgorund process, for example:
```
#!/bin/bash
command_to_execute_in_background &
command1_to_execute_before_connection
command2_to_execute_before_connection
```
So remmina should only take care to wait for script termination before proceeding with the connection.
We can also agree in a kind of return code from the script: if the script/command returns 0, then remmina can proceed to connection. Otherwise Remmina will not connect and show a message to the user.

**Comment by giox069** [comment]
Or better
```
#!/bin/bash
command_to_execute_in_bacground &
commands_to_execute_before_connection
sleep(X) or wait_for_background_command_to_be_active()
exit(0)
```

**Comment by antenore** [comment]
g_app_info_create_from_commandline execute a specified command in foreground, therefore we cannot just check the exit status of the script (or of a program), because just before executing the connection the foreground process could be still running (so no exit status at all).

Your initial proposal, of developing a dialog waiting for the end of the execution, seems to me the most efficient one.

g_app_info_create_from_commandline ---> Wait for end of script/program ---> if 0 --> start connection

**Comment by BrainwreckedTech** [comment]
In preparation for this, I've added an option (`-t`) to `qcl` to allow the user to specify an X terminal to execute `qemu` inside of and in the background.  The -t option adds a 2-second pause (`sleep 2`) to allow `qemu` to get up and running before allowing `qcl` to exit.

Without `-t`, `qcl` will execute `qemu` in the current terminal in the foreground, and thus not exit until QEMU is finished.

TODO: Validate the `term-bin` specified by `-t`, allowing `none` to completely forego the `qemu` console.

**Comment by antenore** [comment]
@BrainwreckedTech 
Can we close this? Did I forget anything else?

**Comment by BrainwreckedTech** [comment]
There's only one minor problem in the implementation.  If the Pre Command executes, forks, and exits, everything works as expected.  If the Pre Command executes in the foreground, Remmina hangs if the VNC connection is used to shut down the remote.

Everything works as expected with this line as the Pre Command:
sudo qcl -w -x 50 -m 512 -p /mnt/qvm/clib-arch.img -t /usr/bin/konsole

Omitting -t causes QCL to execute QEMU itself in the foreground:
sudo qcl -w -x 50 -m 512 -p /mnt/qvm/clib-arch.img

Under the latter condition, PLEASE WAIT appears as the script won't exit until QEMU is done.  Pressing the Cancel button has the intended effect (outside of the PLEASE WAIT dialog remaining) and things work as expected until Remmina's VNC connection is used to shut down the remote machine.

**Comment by giox069** [comment]
Can you try to put your command in a script file like the following, and ask remmina to execute the script:

```
#!/bin/bash
sudo qcl -w -x 50 -m 512 -p /mnt/qvm/clib-arch.img &
sleep 2
```
In this way, the qcl command will go in the backtround ('&' at the end of the line) and the sleep 2 will give 2 seconds to your VM to power on before the script terminate and Remmina will take control...

**Comment by BrainwreckedTech** [comment]
Of course, doing exactly as giox069 pointed out would have the same result as just using QCL with the -t option.  But that gave me an interesting idea.  I used this as the Pre Command:

sudo qcl -w -x 50 -m 512 -p /mnt/qvm/clib-arch.img & sleep 2

The PLEASE WAIT dialog remained indefinitely until Cancel was clicked, but Remmina did not hang when its VNC connection was used to shut down the remote 2 out of 3 times.  _Something_ is going on, there.

Remmina works as expected when the Pre Command works as expected.  I think the trouble now is ironing out how Remmina reacts when the Pre Command does something *not* expected.

**Comment by BrainwreckedTech** [comment]
Also, if there's improperly-quoted text in the Pre Command, Remmina bombs.

** (remmina:24019): WARNING **: Text ended before matching quote was found for '. (The text was '/usr/bin/sudo /scr/qcl -w -x 50 -m 512 -p /mnt/qvm/clib-arch.img'')

** (remmina:24019): WARNING **: (null)

zsh: segmentation fault (core dumped)  remmina

**Comment by antenore** [comment]
I've closed this, solving the segfault, please open new issues for enhancements, bugs, and so forth.
